tools/stylelint/theme-mixin-api.js

```diff
@@ -0,0 +1,113 @@
+const stylelint = require('stylelint');
+
+/** Name of this stylelint rule. */
+const ruleName = 'material/theme-mixin-api';
+
+/** Regular expression that matches all theme mixins. */
+const themeMixinRegex =
+    /^(?:_(mat-.+)-(density)|(mat-.+)-(density|color|typography|theme))\(([^,]+)\)$/;
+
+/**
+ * Stylelint plugin which ensures that theme mixins have a consistent API. Besides
+ * compilation API tests which are stored in `src/material/core/theming/test`, we test
+ * the following patterns here:
+ *
+ *   1. Checks if theme mixin arguments named consistently e.g. if a mixin accepts both a theme
+ *      or color configuration, the variable should reflect that.
+ *   2. Checks if the individual theme mixins handle the case where consumers pass a theme object.
+ *      For convenience, we support passing theme object to the scoped mixins.
+ *   3. Checks if the `-theme` mixins have the duplicate style check set up. We want to
+ *      consistently check for duplicative theme styles so that we can warn consumers.
+ */
+const plugin = stylelint.createPlugin(ruleName, (isEnabled, options, context) => {
+  return (root, result) => {
+    if (!isEnabled)
+      return;
+
+    root.walkAtRules('mixin', node => {
+      const matches = node.params.match(themeMixinRegex);
+      if (matches === null) {
+        return;
+      }
+
+      const componentName = matches[1] || matches[3];
+      const type = matches[2] ||matches[4];
+      const variableName = matches[5];
+
+      if (type === 'theme') {
+        validateThemeMixin(node, componentName, variableName);
+      } else {
+        validateIndividualSystemMixins(node, type, variableName);
+      }
+    });
+
+    function validateThemeMixin(node, componentName, variableName) {
+      if (variableName !== '$theme-or-color-config') {
+        if (context.fix) {
+          node.params = node.params.replace(variableName, '$theme-or-color-config');
+        } else {
+          reportError(node, 'Expected mixin argument to be called `$theme-or-color-config`.');
```

Comment:
"first mixin argument"

Reply:
Yep. Improving that. Now it will visit all potential themes and throw if there are too much arguments. This will ensure that theme mixins always have the same signature.